core/src/main/java/io/aiven/klaw/service/UiControllerLoginService.java

```diff
@@ -52,30 +53,43 @@ public class UiControllerLoginService {
 
   public String getReturningPage(String uri) {
     try {
-      UserDetails userDetails =
-          (UserDetails) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
-      if (userDetails != null) {
+      String userName;
 
+      if ("azuread".equals(authenticationType)) {
+        DefaultOidcUser userDetails =
+            (org.springframework.security.oauth2.core.oidc.user.DefaultOidcUser)
+                SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+        userName = userDetails.getPreferredUsername();
+      } else {
+        UserDetails userDetails =
+            (UserDetails) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+        userName = userDetails.getUsername();
+      }
```

Comment:
In the future we might want to think about having some AD structure if we need to support several different AD flavours. Otherwise this if-else will grow too much.

Reply:
Agree and most of the AD or OAuth mechanisms return DefaultOAuth2User and DefaultOidcUser is child of it.